### PR TITLE
Add error handling for too large JSON payloads

### LINKED ISSUE #972: Return a 413 instead of 400 when data limit is exceeded
Hey there,

when sending a too large JSON payload (i.e. larger than the specified limit) Rocket will show a missleading error message in the debug log and respond with 400 Bad Request.

Instead of this behaviour, I'd expect Rocket to show a error message mentioning the too large payload and respond with 413 Payload Too Large.

Rocket version: 0.4.0
OS: Debian buster (10.0)

## Code:
```
#![feature(proc_macro_hygiene, decl_macro)]
use rocket::*;
use rocket_contrib::json::Json;

#[post("/", data = "<data>")]
fn x(data: Json<String>) {}

fn main() {
    ignite().mount("/", routes![x]).launch();
}
```

```
$ curl -H Content-Type:\ application/json -d @really-large-json-file http://localhost:8000

            <!DOCTYPE html>
            <html>
            <head>
                <meta charset="utf-8">
                <title>400 Bad Request</title>
            </head>
            <body align="center">
                <div align="center">
                    <h1>400: Bad Request</h1>
                    <p>The request could not be understood by the server due
                to malformed syntax.</p>
                    <hr />
                    <small>Rocket</small>
                </div>
            </body>
            </html>
```

```
🔧 Configured for development.
    => address: localhost
    => port: 8000
    => log: normal
    => workers: 8
    => secret key: generated
    => limits: forms = 32KiB
    => keep-alive: 5s
    => tls: disabled
🛰  Mounting /:
    => POST / (x)
🚀 Rocket has launched from http://localhost:8000
POST / application/json:
    => Matched: POST / (x)
    => Warning: Data left unread. Force closing network stream.
    => Error: Couldn't parse JSON body: Error("EOF while parsing a string", line: 1, column: 1048576)
    => Outcome: Failure
    => Warning: Responding with 400 Bad Request catcher.
    => Response succeeded.
```

Best,
Felix

### PR DESCRIPTION
**Pull request by Nilix007**
Hi,

This PR adds error handling for too large JSON payloads. It also adds a new `JsonError` `PayloadTooLarge` so that this case can be handled properly. Fixes #972.

Old behaviour for too large JSON payloads:
```
POST / application/json:
    => Matched: POST / (x)
    => Warning: Data left unread. Force closing network stream.
    => Error: Couldn't parse JSON body: Error("EOF while parsing a string", line: 1, column: 1048576)
    => Outcome: Failure
    => Warning: Responding with 400 Bad Request catcher.
    => Response succeeded.
```

New behaviour:
```
POST / application/json:
    => Matched: POST / (x)
    => Error: JSON payload too large.
    => Warning: Data left unread. Force closing network stream.
    => Outcome: Failure
    => Warning: Responding with 413 Payload Too Large catcher.
    => Response succeeded.
```

### CODE REVIEW

**Comment by jebrosen** [comment]
I like the idea. I think it would be great if Rocket provided an easy way to do this and could share the implementation with `Form` and anything else that has a limit.

File: contrib/lib/src/json.rs [hunk 1]
```diff
@@ -138,7 +141,14 @@ impl<'a, T: Deserialize<'a>> FromData<'a> for Json<T> {
     fn transform(r: &Request, d: Data) -> Transform<Outcome<Self::Owned, Self::Error>> {
         let size_limit = r.limits().get("json").unwrap_or(LIMIT);
         let mut s = String::with_capacity(512);
-        match d.open().take(size_limit).read_to_string(&mut s) {
+        match d.open().take(size_limit + 1).read_to_string(&mut s) {
+            Ok(_) if s.len() as u64 == size_limit + 1 => {
```
Comment (jebrosen) [inline contrib/lib/src/json.rs]: Let's do `if s.len() as u64 > size_limit` - that should make refactoring less brittle.

On that note, the number of bytes read is already returned by `read_to_string` -- `Ok(n) if n as u64 > size_limit` should work as well.

Comment (Nilix007) [reply contrib/lib/src/json.rs]: Oh, yeah, of course :laughing:

I updated the PR.

**Comment by hellow554** [comment]
ping @jebrosen any more concerns? I'd like to see this in rocket!

**Comment by Nilix007** [comment]
I opened #990 which shall provide a general solution for all body data parsers.

Let's continue over there.

**Pull request closed without merging**